### PR TITLE
FFmpeg installs + testing on multiple versions

### PR DESCRIPTION
**Pull request by JamesF2440**
static-ffmpeg now pip installs and is listed in the requirements!
Also added a test, although it is a very simple one
PyTest now runs for python 3.8, 3.8, and 3.10
Section crops, time and frame splits, and now all frame grabs are done via openCV to speed things up.
Tried to package everything up but some of it is still a little weird, may need help on that looking forward.
Committed a test video as well so that anyone cloning the repo won't have to go download a video

### CODE REVIEW

File: aEye/testVid.mp4
Comment (vinayakSharm) [inline aEye/testVid.mp4:1]: hi james, lets read videos from s3. No videos will go on github.

File: aEye/testing.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+if __name__ == "__main__":
+    """
+    Main class to run/test video processing framework. Basically
+    everything should run as a method for the video object, but theres
+    a few limitations: Can't easily multiprocess currently, that's on the horizon for 
+    right now, and some methods made with ffmpeg could get ported over to opencv for
+    the sake of speed. Video join doesnt work currently. 
+    """
+    print("--- Video Processing Framework ---")
+    video = Video("testVid.mp4") # or any local video, creates Video object
```
Comment (vinayakSharm) [inline aEye/testing.py:12]: lets read videos from s3 pls

File: tests/extract_metadata_test.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+from aEye import Video
+
+"""
+trash metadata test class featuring a super epic hardcoded video ref
+"""
+
+
+def test_extract_metadata():
+    new_vid_obj = Video("aEye/testVid.mp4")
```
Comment (vinayakSharm) [inline tests/extract_metadata_test.py:9]: needs to read video from s3.

File: aEye/video.py [hunk 3]
```diff
@@ -1,95 +1,355 @@
...
-            resized = cv2.resize(image , (400,400), interpolation = cv2.INTER_AREA)
+            resized = cv2.resize(image, (400, 400), interpolation=cv2.INTER_AREA)
             self.frame_array.append(resized)
             out.write(resized)
         out.release()
         self.cap.release()
-        
-    def extract_time_frame(self):
+
+    ##############################################
```
Comment (vinayakSharm) [inline aEye/video.py:292]: all code going into main needs to be fully functional. Please move all code that is WIP below into another branch, and keep it for another feature update or next PR

File: aEye/video.py [hunk 4]
```diff
@@ -1,95 +1,355 @@
...
+        Necessary for basically any processing, but still has to be set (none by default)
+    get_codec -> str:
+        Returns the video codec
+    get_duration -> str:
+        Returns video duration in seconds, but does so as a string
+    get_frames -> str:
+        Returns the amount of frames in the video as a string (via OpenCV)
+    getfile -> str:
+        Returns video file path
+    split_on_time(start, end) -> None:
```
Comment (vinayakSharm) [inline aEye/video.py:37]: 37-64 will go in the processor class. 

Lets integrate these as pipeline processing functionalities. You load a video, you process a video, and then save it back to S3.

File: aEye/video.py [hunk 5]
```diff
@@ -1,95 +1,355 @@
...
+    file: str
+        path of the video file associated with the object
+    meta_data: str
+        JSON dictionary of video metadata, typically streams[0] is the video metadata
+    cv_video: cv2.VideoCapture
+        OpenCV video object, used for any openCV processing
+    ----------
+    Methods
+    ----------
+    extract_metadata -> str:
```
Comment (vinayakSharm) [inline aEye/video.py:26]: 26-35 will stay in the video class.

File: aEye/testVid.mp4
Comment (JamesF2440) [reply aEye/testVid.mp4:1]: Got it. This'll be the next thing I look into, just wanted to include something for functionality's sake.

File: tests/extract_metadata_test.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
+from aEye import Video
+
+"""
+trash metadata test class featuring a super epic hardcoded video ref
```
Comment (vinayakSharm) [inline tests/extract_metadata_test.py:4]: haha pls dont say module is trash :). will be really good once we can get it to read from s3!

File: tests/extract_metadata_test.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+"""
+
+
+def test_extract_metadata():
+    new_vid_obj = Video("aEye/testVid.mp4")
+    to_check = new_vid_obj.extract_metadata()
+    codec = to_check["streams"][0]["codec_name"]
+    assert codec == "h264"
+    duration = to_check["streams"][0]["duration"]
+    assert duration == "10.040000"
```
Comment (vinayakSharm) [inline tests/extract_metadata_test.py:14]: good.

File: aEye/video.py [hunk 8]
```diff
@@ -1,95 +1,355 @@
...
+            "crop=720:480:0:0"
+        )
+        with open("filter_script.txt", "w") as file:
+            file.write(script)
+        cmd = (
+            f"{ffmpeg} -y -v quiet -ss 0 -t 30 -i {self.getfile()} -filter_complex_script filter_script.txt -acodec copy "
+            f"outputs/moving.mp4"
+        )
+        subprocess.call(cmd, shell=True)
+        os.remove("filter_script.txt")
```
Comment (vinayakSharm) [inline aEye/video.py:225]: nice!

Question: do we need to write this script to a txt file or can we run it directly?

File: aEye/video.py [hunk 9]
```diff
@@ -1,95 +1,355 @@
...
+        -------
+        None, but creates trimmed video in output folder
+        """
+        fps = float(self.get_frames()) / float(self.get_duration())
+        time_stamp = start_frame / fps
+        cmd = (
+            f"{ffmpeg} -y -ss {str(time_stamp)} -i {self.getfile()} -v quiet -c:v libx264 -frames:v {num_frames}"
+            f" outputs/frames_{start_frame}_to{(start_frame+num_frames)}.mp4"
+        )
+        print("Encoding ", num_frames, " from ", start_frame)
```
Comment (vinayakSharm) [inline aEye/video.py:181]: lets use logging.info

File: aEye/video.py [hunk 10]
```diff
@@ -1,95 +1,355 @@
...
+
+    def get_frames(self):
+        """
+        Returns # of frames in video as a string
+        """
+        return self.cv_video.get(cv2.CAP_PROP_FPS)
+
+    def getfile(self):
+        """
+        returns da file path as a string
```
Comment (vinayakSharm) [inline aEye/video.py:115]: *the

File: aEye/video.py [hunk 11]
```diff
@@ -1,95 +1,355 @@
...
+        String of JSON Dictionary full of video metadata
+        """
+        command = f"{ffprobe} -hide_banner -show_streams -v error -print_format json -show_format -i '{self.getfile()}'"
+        out = subprocess.check_output(command, shell=True).decode("utf-8")
+        json_data = json.loads(out)
+        return json_data
+
+    def get_codec(self):
+        """
+        Returns video codec as a string
```
Comment (vinayakSharm) [inline aEye/video.py:89]: full sentences pl

**Review by vinayakSharm — CHANGES_REQUESTED** [review]
Hi james some quick feedback:

(1) Video class is looking solid!
(a) keep the functions mentioned in video class
(b) move the other functions in the processor class that should be in main now

(2) No uploading a movie to github. 22mb is far too large. 

(3) read and write everything to s3 or from s3. Use the load and upload functionality from the processor class. 

(4) Comments need to be pro. so close :). 

(5) remove all code that does not work right now. save it for a later feature update to the code base. 

(6) replace all print statements with logging.

File: aEye/processor.py [hunk 12]
```diff
@@ -63,29 +106,27 @@ def load(self, bucket=  'aeye-data-bucket', prefix='input_video/') -> list[Video
...
+            # This if-statement is to skip over the folder object since we are only interested in the video files.
             if i["Key"] == prefix:
                 continue
 
             title = i["Key"].split(prefix)[1]
-            #In order to convert video file from S3 to cv2 video class, we need its url.
-            url = s3.generate_presigned_url( ClientMethod='get_object', Params={ 'Bucket': bucket, 'Key': i["Key"] } ,ExpiresIn=5)
+            # In order to convert video file from S3 to cv2 video class, we need its url.
+            url = self._s3.generate_presigned_url(ClientMethod='get_object', Params={'Bucket': bucket, 'Key': i["Key"]},
+                                                  ExpiresIn=100000)
```
Comment (luanlam97) [inline aEye/processor.py:121]: expiration time is too long, makes it shorter pls

File: aEye/processor.py [hunk 13]
```diff
@@ -172,6 +210,167 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
         print("successfully upload the output files S3 bucket: s3://aeye-data-bucket/modified/")
         print("successfully remove the output file from local machine")
 
+    def split_on_time(self, start, end):
```
Comment (luanlam97) [inline aEye/processor.py:213]: change function to be english alike
ex: trim_video_from_to( start, end)

File: aEye/processor.py [hunk 14]
```diff
@@ -172,6 +210,167 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
+        -------
+        None, but creates trimmed video in output folder
+        """
+        for video in self.video_list:
+            duration = end - start
+            cmd = f"{ffmpeg} -y -ss {start} -i '{video.getfile()}' -v quiet -t {duration} -c copy " \
+                  f"modified/split_on_time_{video.title}"
+            subprocess.call(cmd, shell=True)
 
+    def split_on_interval(self, interval):
```
Comment (luanlam97) [inline aEye/processor.py:228]: change this as well
ex: divide_video_by_secs

File: aEye/processor.py [hunk 15]
```diff
@@ -172,6 +210,167 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
+            fps = float(video.get_frames()) / float(video.get_duration())
+            time_stamp = start_frame / fps
+            cmd = (
+                f"{ffmpeg} -y -ss {str(time_stamp)} -i '{video.getfile()}' -v quiet -c:v libx264 -frames:v {num_frames}"
+                f" modified/frames_{start_frame}_to{(start_frame + num_frames)}_{video.title}.mp4"
+            )
+            logging.info("Encoding ", num_frames, " from ", start_frame)
+            subprocess.call(cmd, shell=True)
+
+    def crop_video_section(self, section_width, section_height, start_x, start_y):
```
Comment (luanlam97) [inline aEye/processor.py:285]: switch and make the start_x and start_y as the first part of the parameter

File: aEye/video.py [hunk 16]
```diff
@@ -1,80 +1,100 @@
...
+        """
+        if self.meta_data is not None:
+            return self.meta_data["streams"][0]["duration"]
+        else:
+            self.meta_data = self.get_metadata()
+            return self.meta_data["streams"][0]["duration"]
 
-            
-    def __repr__(self):
+    def get_frames(self):
```
Comment (luanlam97) [inline aEye/video.py:84]: change to get_num_frames()

File: aEye/video.py [hunk 17]
```diff
@@ -1,80 +1,100 @@
...
-        return self.title
-            
+        if self.meta_data is not None:
+            return self.meta_data["streams"][0]["nb_frames"]
+        else:
+            self.meta_data = self.extract_metadata()
+            return self.meta_data["streams"][0]["nb_frames"]
 
-    def cleanup(self) -> None:
+    def getfile(self):
```
Comment (luanlam97) [inline aEye/video.py:94]: change to get_file

**Pull request closed without merging**